src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,100 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        elementData = growIfArrayFull();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        elementData = growIfArrayFull();
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index
+                    + ". Your index must be in the range from 0 to " + size + " inclusive.");
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        elementData = growIfArrayFull();
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedElement = get(index);
+        System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((elementData[i] != null && elementData[i].equals(element))
+                    || elementData[i] == element) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("There is no such element present in the list.");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T[] growIfArrayFull() {
+        T[] grownArray = elementData;
+        int oldCapacity = elementData.length;
+        if (size == oldCapacity) {
+            int newCapacity = oldCapacity + (oldCapacity >> 1);
+            grownArray = (T[]) new Object[newCapacity];
+            System.arraycopy(elementData, 0, grownArray, 0, size);
+        }
+        return grownArray;
+    }
```

Comment:
1. Matter of taste but for me it seems that this method is more understandable when void as we always update class variable that is accessible from method.
2. Code seems more readable when we have validations first and then main code part which not covered by if block
```suggestion
    private void growIfArrayFull() {
        if (size < oldCapacity) {
            return;
        }
        int newCapacity = elementData.length + (elementData.length >> 1);
        grownArray = (T[]) new Object[newCapacity];
        System.arraycopy(elementData, 0, grownArray, 0, size);
        elementData = grownArray;
    }
```